### ISSUE TITLE
Composer occasionally generates bogus JSON for dialogs

### ISSUE THREAD

**Issue by Stevenic** [microsoft/BotFramework-Composer#3657]
## Describe the bug
Composer occasionally generates bogus JSON for dialogs.  This seems to have started with 1.0.2 and I've hit it twice now. Both cases were the same with the last few bytes of the file being duplicated.

## Version
1.0.2

## Screenshots

![image](https://user-images.githubusercontent.com/698939/87861309-dee23500-c8f9-11ea-8c16-3c12a1d2d5ec.png)

**Comment by cwhitten** [comment]
This looks like a race condition when serializing. Tagging @lei9444

**Comment by lei9444** [comment]
I will investigate the issue

**Comment by beyackle** [comment]
What I've found so far: the actual file-saving thing happens in the server, at server/src/models/bot/botProject.ts. It looks like there's no lock in place to make sure we're not trying to save the file twice at the same time.

**Comment by lei9444** [comment]
> What I've found so far: the actual file-saving thing happens in the server, at server/src/models/bot/botProject.ts. It looks like there's no lock in place to make sure we're not trying to save the file twice at the same time.

Yes, this is from server. I can't reproduce it. I think this issue is caused by like this
```
const  fs = require('fs');
const { promisify } = require('util')

const writeFileAsync = promisify(fs.writeFile)

const textPath = __dirname + '\\test.txt';

const run = async () => {

  await writeFileAsync(textPath, `a
  b
  c`)
}
 

const run1 = async () => {
  await writeFileAsync(textPath, `xx`)
}

run()
run1()
```

**Comment by beyackle** [comment]
@lei9444 That's what I concluded too. This is hard to repro, but there might be a workaround - if it's the last section of the file getting written twice somehow, then what if we just debounced the write operation? Make it so writing/updating a file can't happen twice within a certain interval.

**Comment by cwhitten** [comment]
@Stevenic is saying he is running into this multiple times per day. Steve, can you give us a consistent repro in the context of Composer?

**Comment by cwhitten** [comment]
@beyackle I like the thought. If we had a consistent repro from @Stevenic and applied the workaround to it I'd be open to take your suggestion.

**Comment by beyackle** [comment]
I've tried to repro it by throttling the connection speed down to "Slow 3G" and leaving the ".dialog" file open while I edit a bot, and the changes just propagate into the file cleanly with no issue. Actual repro steps would be really useful here.
